### PR TITLE
Add object tagging - Moodle 42

### PR DESCRIPTION
**Pull request by matthewhilton**
- Cherry pick of https://github.com/catalyst/moodle-tool_objectfs/pull/619 to `MOODLE_402_STABLE` branch
- Updated version.php and upgrade.php version numbers to todays date since this is the newest branch.

### CODE REVIEW

**Comment by matthewhilton** [comment]
Leaving as draft until https://github.com/catalyst/moodle-tool_objectfs/pull/619 is in a mergeable-state

File: classes/task/update_object_tags.php [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        // For each, try to sync their tags.
+        foreach ($contenthashes as $contenthash) {
+            $fs->push_object_tags($contenthash);
+        }
+
+        // Re-queue self to process more in another iteration.
+        mtrace("Requeing self for another iteration.");
+        $task = new update_object_tags();
+        $task->set_custom_data([
+            'iteration' => $iteration + 1,
```
Comment (brendanheywood) [inline classes/task/update_object_tags.php:120]: I don't understand why we'd want these as adhoc tasks with customdata over a scheduled task?

Comment (matthewhilton) [reply classes/task/update_object_tags.php:120]: AFAIK scheduled tasks don't have super nice retry logic and if they are long running will block the queue of other scheduled tasks, whereas adhoc tasks are a lot more robust and parallelise better

**Comment by matthewhilton** [comment]
Rebased, but marking as draft again since the 310 PR has made some changes that i'll need to cherry pick over

**Comment by matthewhilton** [comment]
Just updated this PR:
- Cherry picked the squashed commit from https://github.com/catalyst/moodle-tool_objectfs/pull/619
- Completed manual testing including upgrading - works as expected
- Ran phpunit tests with the s3 integration client  - no issues other than expected perms issue (e.g. trying to test delete but token has no delete perms)

Once the CI is good, i'll merge this in.

**Pull request merged**